scripts/db-heal-test.py

```diff
@@ -0,0 +1,284 @@
+def has_program_failed(code, log):
+    if code != 0:
+        log.close()
+        with open(carmen_log_file, 'r') as l:
+            text = l.read()
+            print(text)
+        return True
+    return False
+
+
+# Function which checks every line added to aida_log_file and behaves accordingly to the line.
+def check_aida_log(sleep_time: int, checkpoint: int):
+    start = 0.0
+    with open(aida_log_file, 'r') as f:
+        while True:
+            line = f.readline()
+            if not line:
+                time.sleep(0.1)
+                continue
+            if start > 0 and time.time() - start >= sleep_time:
+                print("Interrupting...")
+                process.terminate()
+                return False
+            if start == 0.0 and "Track: block" in line:
+                # This line in combination with 0 start time means Syncing has been restarted.
+                print("Syncing restarted!")
+                start = time.time()
+            # If process ends with error (return code 1) or either 'fail' or 'exit status' occurs in line exit script
+            if process.poll() == 1 or any(s in line for s in ["exit status", "fail"]):
+                print("Error occurred - printing output.log:")
+                with open(aida_log_file, 'r') as l:
+                    text = l.read()
+                    print(text)
+                return True
+
+
+# Function which runs Carmen's info command and finds the latest checkpoint from created log
+def get_latest_checkpoint_from_info():
+    log = os.path.join(working_dir, 'carmen-info.log')
+    cp: str
+    with open(log, 'w') as cl:
+        r = subprocess.run(
+            ['go', 'run', './database/mpt/tool', 'info', str(archive)],
+            stdout=cl,
+            stderr=cl,
+            cwd=carmen_root)
+        if has_program_failed(r.returncode, cl):
+            return -1
+
+    with open(log, 'r') as cl:
+        info_checkpoint = cl.readlines()[-1]
+        # Return last word which is the block number
+        cp = info_checkpoint.split()[-1]
+
+    os.remove(log)
+    return int(cp)
+
+
+# Function which finds newest dir within path p hence finds the current working db.
+def find_working_db(p):
+    result = []
+    for d in os.listdir(p):
+        pd = os.path.join(p, d)
+        if os.path.isdir(pd):
+            result.append(pd)
+
+    return max(result, key=os.path.getmtime)
+
+
+# First iteration command
+cmd = [
+    './build/aida-vm-sdb', 'substate', '--validate',
+    '--db-tmp', working_dir, '--carmen-schema', '5', '--db-impl', 'carmen',
+    '--aida-db', aida_db_path, '--no-heartbeat-logging', '--track-progress',
+    '--archive', '--archive-variant', 's5', '--archive-query-rate', '200',
+    '--carmen-checkpoint-interval', str(checkpoint_granularity), '--tracker-granularity',
+    str(checkpoint_granularity), str(first_block), str(last_block)
+]
+
+os.chdir(aida_path)
```

Comment:
do you really need to `chdir` here and there? It looks to me that all paths are absolute to the temp dir.

Reply:
also holds for other places in the script